Introduce DCN DT

sbekkerm opened the pull request:
This PR introduces DCN VA, which builds upon the HCI VA architecture and is designed for multi-site deployment.

In addition to the regular configuration files, this PR includes Jinja templates for generating values.yaml and service-values.yaml files. These templates are essential for Zuul job execution, allowing for the creation of site-specific configuration files multiple times for each DCN site.

abays commented on examples/va/dcn/README.md:
```diff
@@ -0,0 +1,42 @@
+# Distributed Compute Node (DCN) OpenStack Architecture with HCI and Ceph
+
+**Based on OpenStack K8S operators from the "main" branch of the [OpenStack Operator repo](https://github.com/openstack-k8s-operators/openstack-operator/commit/aa63bf3931f74722dd48af8a0914233b2b384330) on Dec 19th, 2023**
```
Please update this to indicate which commit of OpenStack operator you used to test this.

fultonj replied:
I had a good run of this using the following. I could ask for Sergey to put that in the README with today's date. 

https://github.com/openstack-k8s-operators/openstack-operator/commit/ec7210b825a3c355007d1f1fc11a2952ba4a9262

fultonj replied:
It has been added to latest update.

abays commented on automation/vars/dcn.yaml:
This seems to be missing stages for the data plane.  Or was that intentional?

sbekkerm replied:
I removed the data plane deployment to avoid issues with the Zuul jobs. The deployment is handled by a custom ansible playbook since, for DCN, we need to repeat Steps 3, 4, and install Ceph for each DCN site. This PR has been tested within downstream, job #665

fultonj replied:
Do you want to put that playbook in a ci-framwork patch and document that it's something to be run separately?

  https://github.com/openstack-k8s-operators/ci-framework/tree/main/playbooks

Have a look at this:

  https://github.com/openstack-k8s-operators/architecture/blob/main/automation/vars/hci.yaml#L50-L54

In theory, you could add a `post_stage_run` in this PR which calls your new playbook and have a depends-on.

What do you think @cjeanner ?

fultonj replied:
Yes, the post_stage run is now used and these concerns are addressed.

fultonj commented:
@sbekkerm I see two changes needed up front.

1. Written Instructions

The readme files are incomplete. Please see the the four stage readme's for VA HCI:

  https://github.com/openstack-k8s-operators/architecture/tree/main/examples/va/hci#stages

It contains English instructions that someone can read to implement the VA without ci-framework and using only the produced k8s manifests. If there are external automations for them, that's fine but I should be able to read the directions and reproduce your work so that we can have independent verification. Right now it looks like the the VA1 directions are still there and not updated. In my early example someone could read my directions and get a full deployment (and the `extra` directory with scripts can technically be ignored).

  https://github.com/fultonj/dcn?tab=readme-ov-file#steps

No code should be required to implement what I'm talking about for this request. Just written instructions.

2. VA vs DT

Would you please change this so that it puts the added files into the `dt` directory instead of the `va` directory?

sbekkerm commented:
> @sbekkerm I see two changes needed up front.
> 
> 1. Written Instructions
> 
> The readme files are incomplete. Please see the the four stage readme's for VA HCI:
> 
> https://github.com/openstack-k8s-operators/architecture/tree/main/examples/va/hci#stages
> 
> It contains English instructions that someone can read to implement the VA without ci-framework and using only the produced k8s manifests. If there are external automations for them, that's fine but I should be able to read the directions and reproduce your work so that we can have independent verification. Right now it looks like the the VA1 directions are still there and not updated. In my early example someone could read my directions and get a full deployment (and the `extra` directory with scripts can technically be ignored).
> 
> https://github.com/fultonj/dcn?tab=readme-ov-file#steps
> 
> No code should be required to implement what I'm talking about for this request. Just written instructions.
> 
> 2. VA vs DT
> 
> Would you please change this so that it puts the added files into the `dt` directory instead of the `va` directory?

1. It contains the instructions. All four DCN steps are almost the same as HCI VA, except for the post-nova actions, which are already covered here: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-post-ceph.md#finalize-nova-computes
Additionally, it mentions that Steps 3, 4, and the Ceph installation need to be executed for each DCN site

The main difference between VA and HCI is in the values.yaml and service-values.yaml files. For example, the nncp values.yaml contains the configuration necessary for spine and leaf: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/control-plane/nncp/values.yaml#L18

and the post-ceph service-values.yaml contains Glance Multi Store configuration:
https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/service-values.yaml#L117


2. Why should we use DT instead of VA?

fultonj commented:
> > @sbekkerm I see two changes needed up front.
> > 
> > 1. Written Instructions
> > 
> > The readme files are incomplete. Please see the the four stage readme's for VA HCI:
> > https://github.com/openstack-k8s-operators/architecture/tree/main/examples/va/hci#stages
> > It contains English instructions that someone can read to implement the VA without ci-framework and using only the produced k8s manifests. If there are external automations for them, that's fine but I should be able to read the directions and reproduce your work so that we can have independent verification. Right now it looks like the the VA1 directions are still there and not updated. In my early example someone could read my directions and get a full deployment (and the `extra` directory with scripts can technically be ignored).
> > https://github.com/fultonj/dcn?tab=readme-ov-file#steps
> > No code should be required to implement what I'm talking about for this request. Just written instructions.
> > 
> > 2. VA vs DT
> > 
> > Would you please change this so that it puts the added files into the `dt` directory instead of the `va` directory?
> 
> 1. It contains the instructions. All four DCN steps are almost the same as HCI VA, except for the post-nova actions, which are already covered here: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-post-ceph.md#finalize-nova-computes
>    Additionally, it mentions that Steps 3, 4, and the Ceph installation need to be executed for each DCN site
> 
> The main difference between VA and HCI is in the values.yaml and service-values.yaml files. For example, the nncp values.yaml contains the configuration necessary for spine and leaf: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/control-plane/nncp/values.yaml#L18
> 
> and the post-ceph service-values.yaml contains Glance Multi Store configuration: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/service-values.yaml#L117

Yes, I understand but I still need clear written instructions so I can reproduce. 

> 2. Why DT?

https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/README.md

I don't think this should be an update to an existing DT, it should be a new DT, but it's a DT, not a VA.

I wouldn't want to hand the field the Jinja2 files. This is something we do for our CI but not yet ready to be a full blown VA we could hand to someone in the field. Maybe it can evolve into a VA in the future. For now, in order to merge what you have I think it should be a DT.

sbekkerm commented:
> > > @sbekkerm I see two changes needed up front.
> > > 
> > > 1. Written Instructions
> > > 
> > > The readme files are incomplete. Please see the the four stage readme's for VA HCI:
> > > https://github.com/openstack-k8s-operators/architecture/tree/main/examples/va/hci#stages
> > > It contains English instructions that someone can read to implement the VA without ci-framework and using only the produced k8s manifests. If there are external automations for them, that's fine but I should be able to read the directions and reproduce your work so that we can have independent verification. Right now it looks like the the VA1 directions are still there and not updated. In my early example someone could read my directions and get a full deployment (and the `extra` directory with scripts can technically be ignored).
> > > https://github.com/fultonj/dcn?tab=readme-ov-file#steps
> > > No code should be required to implement what I'm talking about for this request. Just written instructions.
> > > 
> > > 2. VA vs DT
> > > 
> > > Would you please change this so that it puts the added files into the `dt` directory instead of the `va` directory?
> > 
> > 
> > 
> > 1. It contains the instructions. All four DCN steps are almost the same as HCI VA, except for the post-nova actions, which are already covered here: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-post-ceph.md#finalize-nova-computes
> >    Additionally, it mentions that Steps 3, 4, and the Ceph installation need to be executed for each DCN site
> > 
> > The main difference between VA and HCI is in the values.yaml and service-values.yaml files. For example, the nncp values.yaml contains the configuration necessary for spine and leaf: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/control-plane/nncp/values.yaml#L18
> > and the post-ceph service-values.yaml contains Glance Multi Store configuration: https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/service-values.yaml#L117
> 
> Yes, I understand but I still need clear written instructions so I (or anyone else) can reproduce.
> 
> I want to collaborate with you on this and reproduce the results in my environment so I can find and fix bugs. I think the READMEs are missing too much and filling them in will help other engineers and the docs team. From a very high level it's so we can have https://en.wikipedia.org/wiki/Reproducibility
> 
> > 2. Why DT?
> 
> https://github.com/openstack-k8s-operators/architecture/blob/main/examples/dt/README.md
> 
> I don't think this should be an update to an existing DT, it should be a new DT, but it's a DT, not a VA.
> 
> I wouldn't want to hand the field the Jinja2 files. This is something we do for our CI but not yet ready to be a full blown VA we could hand to someone in the field. Maybe it can evolve into a VA in the future. For now, in order to merge what you have I think it should be a DT.

The README contains all the steps to reproduce the environment. Could you please clarify what specifically is unclear?

1. These steps for deploying the control plane:
https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/control-plane.md
2.  These steps for preparing nodes for ceph installation:
https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-pre-ceph.md
3. These steps for configuring nodes after ceph installation:
https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-post-ceph.md

The Jinja templates are not involved in the deployment process, as they are only used by CI to set the "CHANGEME" parameters in values.yaml and service-valiues.yaml

fultonj commented:
@sbekkerm 

Please move this from the `va` directory to the `dt` directory. 

I am putting it in my backlog to go through your README line by line and attempt to reproduce what you have deployed and I'll ask you clarifying questions along the way which will point out what is incomplete in the READMEs.

sbekkerm commented:
@fultonj 

Moved from `va` to `dt` directory as requested.
Also added a high-level diagram to the README for more clarity. Let me know if you have any questions.

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,246 @@
+      customServiceConfig: |
+        [DEFAULT]
+        enabled_backends = ceph
+        glance_api_servers = http://glance-az2-internal.openstack.svc:9292
```
The TLS is enabled by default so we should use https endpoint:
https://glance-az2-internal.openstack.svc:9292

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,157 @@
+        [DEFAULT]
+        enabled_backends = ceph
+{% if 'ceph' not in _ceph.cifmw_ceph_client_cluster %}
+        glance_api_servers = http://glance-{{ _ceph.cifmw_ceph_client_cluster }}-internal.openstack.svc:9292
```
The TLS is enabled by default so we should use https endpoint:
https://glance-{{ _ceph.cifmw_ceph_client_cluster }}-internal.openstack.svc:9292

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,246 @@
+  cinderAPI:
+    replicas: 3
+  cinderBackup:
+    replicas: 3
```
I'd disable cinder-backup by setting replicas to 0 for now because tempest does have a capability to create a backup in a different/specified AZ anyway

sbekkerm replied:
I would keep the replica as is because it's also a reference architecture. I would skip unsupported tests while they don't support it.

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,246 @@
+data:
+  preserveJobs: false
+  cinderAPI:
+    replicas: 3
```
We need to define the default AZ so some Volume resources such as Volumes groups created in tempest tests have a proper AZ to be created at, something like:
```
cinderAPI:
    replicas: 3
    customServiceConfig: |
      [DEFAULT]
        default_availability_zone = ceph
```

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,246 @@
+            [neutron]
+            physnets = datacentre,leaf1,leaf2
+  nova:
+      customServiceConfig: |
+        [DEFAULT]
+        default_schedule_zone=az-dcn0
```
I can't see the default AZ set anywhere on deployed env and default AZ for nova should be az0 If It's done the way (to be consistent with other parts of generated CRs). This should imo be:
```
nova:
    template:
      apiServiceTemplate:
        customServiceConfig: |
          [DEFAULT]
          default_schedule_zone = az0
```

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/values.yaml.j2:
```diff
@@ -0,0 +1,51 @@
+                images_rbd_glance_copy_poll_interval=15
+                images_rbd_glance_copy_timeout=600
+                rbd_user=openstack
+                rbd_secret_uuid={{ cifmw_ceph_client_fsid }}
```
We should add some more parameters to be set, the glance endpoint and the cross_az_attach parameter, so It could look like:
```

    [glance]
    endpoint_override = https://glance-{{ _az }}-internal.openstack.svc:9292
    valid_interfaces = internal
    [cinder]
    cross_az_attach = False
    catalog_info = volumev3:cinderv3:internalURL
```
Additionally to the [libvirt] section

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/values.yaml.j2:
```diff
@@ -0,0 +1,51 @@
+                images_type=rbd
+                images_rbd_pool=vms
+                images_rbd_ceph_conf=/etc/ceph/{{ cifmw_ceph_client_cluster }}.conf
+                images_rbd_glance_store_name=default_backend
```
I believe It should be:
`                images_rbd_glance_store_name={{ cifmw_ceph_client_cluster }}`

For example for az1:
 images_rbd_glance_store_name=az1

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,157 @@
+            network_vlan_ranges = datacentre:1:1000,leaf1:1:1000,leaf2:1:1000
+            [neutron]
+            physnets = datacentre,leaf1,leaf2
+  nova:
+      customServiceConfig: |
+        [DEFAULT]
+        default_schedule_zone=az0
```
I can't see the default AZ set anywhere on deployed env and default AZ for nova should be az0 If It's done the way (to be consistent with other parts of generated CRs). This should imo be
```
nova:
    template:
      apiServiceTemplate:
        customServiceConfig: |
          [DEFAULT]
          default_schedule_zone = az0
```

sbekkerm replied:
Done

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,157 @@
+  cinderAPI:
+    replicas: 3
+  cinderBackup:
+    replicas: 3
```
I'd disable cinder-backup by setting replicas to 0 for now because tempest does have a capability to create a backup in a different/specified AZ anyway

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,157 @@
+    config.kubernetes.io/local-config: "true"
+data:
+  preserveJobs: false
+  cinderAPI:
```
We need to define the default AZ so some Volume resources such as Volumes groups created in tempest tests have a proper AZ to be created at, something like:

```
cinderAPI:
    replicas: 3
    customServiceConfig: |
      [DEFAULT]
       default_availability_zone = ceph
```

sbekkerm replied:
Done

krcmarik reviewed (commented):
I am suggesting some changes based on what we had in 17.1 openstack services configs and/or to make tempest tests pass (I've managed to make compute and volume tempest suites to pass all tests with the proposed changes which I applied manually)

sbekkerm commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,246 @@
+      customServiceConfig: |
+        [DEFAULT]
+        enabled_backends = ceph
+        glance_api_servers = http://glance-az1-internal.openstack.svc:9292
```
Done

fultonj commented:
@sbekkerm  Could this be rebased?

On Thursday Oct 3rd I'll do a deployment to test this proposed patch and leave feedback (sorry for the delay).

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,249 @@
+          store_description = "az2 RBD backend"
+          rbd_store_pool = images
+          rbd_store_user = openstack
+          rbd_thin_provisioning = True
```
Could we add lines (after this one) to enable automatic image format conversion to raw format? It'd make it easier to import the image in raw format to all glance stores without a need to convert the cirros image from qcow2 to raw format with qemu-img tool (which is not available in opentstackclient pod and It makes it more comfortable if once wanna execute for example tempest test since tempestconf has no ability to create an image in multiple glance stores). It makes sense anyway since we use ceph storage as the backend.
The config lines to enable it would be:
```
            [image_import_opts]
            image_import_plugins = ['image_conversion']
            [image_conversion]
            output_format = raw
```

sbekkerm replied:
done

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,160 @@
+          store_description = "{{ _ceph_az.cifmw_ceph_client_cluster }} RBD backend"
+          rbd_store_pool = images
+          rbd_store_user = openstack
+          rbd_thin_provisioning = True
```
Could we add lines (after this one) to enable automatic image format conversion to raw format? It'd make it easier to import the image in raw format to all glance stores without a need to convert the cirros image from qcow2 to raw format with qemu-img tool (which is not available in opentstackclient pod and It makes it more comfortable if once wanna execute for example tempest test since tempestconf has no ability to create an image in multiple glance stores). It makes sense anyway since we use ceph storage as the backend.
The config lines to enable it would be:

```
            [image_import_opts]
            image_import_plugins = ['image_conversion']
            [image_conversion]
            output_format = raw

```

sbekkerm replied:
done

krcmarik reviewed (commented):
One more suggestion for the CR

sbekkerm commented:
> @sbekkerm @fultonj One more question, I am not sure what the final decision was but do we want to include multicell (cell per site) into this DCN DT? It's just adding some extra configuration for the control plane which I can do.

In my view, it's a great suggestion. This feature will be beneficial in cases where AZ sites contain many compute nodes. Additionally, we can use this when we spread the control plane across AZs.

fultonj commented:
> > @sbekkerm @fultonj One more question, I am not sure what the final decision was but do we want to include multicell (cell per site) into this DCN DT? It's just adding some extra configuration for the control plane which I can do.
> 
> In my view, it's a great suggestion. This feature will be beneficial in cases where AZ sites contain many compute nodes. Additionally, we can use this when we spread the control plane across AZs.

I agree.

My understanding is that there is agreement and that we can go ahead with using multicell in this DT. 

That said there might be people who don't use multicell but I still think we should test this with multicell. If there was a way to run the resultant job with our without multicell by flipping one switch that would be really nice but that might be too much to ask in this first edition.

fultonj commented:
> @sbekkerm @fultonj One more question, I am not sure what the final decision was but do we want to include multicell (cell per site) into this DCN DT? It's just adding some extra configuration for the control plane which I can do.

@krcmarik do you want to try that in a stacked patch.

i.e. create a new PR which has been rebased on this one so that we can merge this one sooner.

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+        rbd_flatten_volume_from_snapshot = False
+        rbd_secret_uuid = CHANGEME
+        rbd_cluster_name = ceph
+        backend_availability_zone = ceph
```
Based on the offline discussion, let's change this to az0 which would be the name unified with compute Availability zone of the default site's compute nodes where this ceph cluster is colocated.
The name of volume and compute AZ must be the same otherwise a similar ERROR will appear:
2024-10-07 09:31:53.192 2 ERROR nova.compute.manager [None req-b437ddf9-b83b-421e-ae44-2756d6392706 958be7b3096d42f28e2c35e6752e5dcb e308ae3b29eb4b9194cf5bbd10536d30 - - default default] [instance: 00528f0a-4fc2-4b06-8aa5-357c504da23c] Failed to attach 6e6af3a3-be24-4c67-9ba8-7b069b1bde7c at /dev/vdb: nova.exception.InvalidVolume: Invalid volume: Instance 00528f0a-4fc2-4b06-8aa5-357c504da23c and volume 6e6af3a3-be24-4c67-9ba8-7b069b1bde7c are not in the same availability_zone. Instance is in az0. Volume is in ceph

fultonj replied:
This has been addressed.

krcmarik commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+    replicas: 3
+    customServiceConfig: |
+      [DEFAULT]
+        default_availability_zone = ceph
```
Let's change this to az0 to align it with the name of compute nova AZ

fultonj replied:
this has been addressed.

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,164 @@
+    replicas: 3
+    customServiceConfig: |
+      [DEFAULT]
+      default_availability_zone = ceph
```
Let's change it to az0

krcmarik commented on examples/dt/dcn/service-values.yaml.j2:
```diff
@@ -0,0 +1,164 @@
+        rbd_flatten_volume_from_snapshot = False
+        rbd_secret_uuid = {{ _ceph.cifmw_ceph_client_fsid }}
+        rbd_cluster_name = {{ _ceph.cifmw_ceph_client_cluster }}
+        backend_availability_zone = {{ _ceph.cifmw_ceph_client_cluster }}
```
Not sure how to do it here so that the value for default site would be az0 and not ceph (which is the volume_backend_name). Would {{ _az }} work?

fultonj replied:
This has been addressed elsewhere.

hjensas commented:
I am not a fan of the introduction of jinja2 templates - and the custom playbook to execute them.
Can we not instead define multiple nodesets, and use the "standard" kustomize_deploy role to apply this?
There are atleast two other PR's proposed that use multiple nodeset approach.
* https://github.com/openstack-k8s-operators/architecture/pull/401
* https://github.com/openstack-k8s-operators/architecture/pull/413

sbekkerm commented:
> I am not a fan of the introduction of jinja2 templates - and the custom playbook to execute them. Can we not instead define multiple nodesets, and use the "standard" kustomize_deploy role to apply this? There are atleast two other PR's proposed that use multiple nodeset approach.
> 
> * [Add multi-cell DT #401](https://github.com/openstack-k8s-operators/architecture/pull/401)
> * [DT - BMO deploy with preprovisioningNetworkData #413](https://github.com/openstack-k8s-operators/architecture/pull/413)
@hjensas
I get what you mean about preferring to use multiple nodesets "standard" role, but the DCN setup is a bit more complex than that.

It’s not just about handling multiple nodesets with spine and leaf—it’s about managing multiple sites, each with its own configuration for services like Glance and Cinder. Plus, each site needs its own Ceph cluster, which adds another layer of complexity. That’s why we went with Jinja2 templates and a custom playbook for this.

But I do agree with you that eventually integrating this into the ci-framework would be ideal

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+          rbd_store_pool = images
+          rbd_store_user = openstack
+          rbd_thin_provisioning = True
+          [image_import_opts]
```
If Ceph is set as a backend, no need to explicitly enable `image_conversion` as the operator does that for you.
Remove L149-L152

krcmarik replied:
I'd need to test it again but I do not think It did, that's why we enabled it.

krcmarik replied:
I understood what you meant, I just believed It did not work - glance-operator did not set it so I tested it and you are right, I believed It did not work most likley because when an image was created no --import cmdline option was specified so the import method was not used for the image creation and that's why the conversion was not triggered. @sbekkerm sorry about that, we cna remove the image conversion lines.

fultonj replied:
I've confirmed that there are no `image_import_opts` set

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+          rbd_store_pool = images
+          rbd_store_user = openstack
+          rbd_thin_provisioning = True
+          [image_import_opts]
```
As per my previous comment, remove L192-L195

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+        replicas: 3
+        type: edge
+  manila:
+    enabled: true
```
disable manila as it's not part of FR1 and probably in Ceph we don't enable `File` (no MDS is deployed, hence this won't work).
Change `enabled: true` to `enabled: false`.
We can keep the config for future Manila deployments if native_cephfs is the target.

fultonj replied:
I've confirmed that manila is disabled in the latest

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+          rbd_store_pool = images
+          rbd_store_user = openstack
+          rbd_thin_provisioning = True
+          [image_import_opts]
```
rm L106-L109 as per my other comments.

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+          rbd_store_pool = images
+          rbd_store_user = openstack
+          rbd_thin_provisioning = True
+          [az0]
```
I assume `az2` only needs az0 (central) and az2 (the local ceph cluster), and the other backends can be removed.

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,261 @@
+          enabled_backends = az0:rbd,az1:rbd,az2:rbd
+          [glance_store]
+          default_backend = az1
+          [az1]
```
I assume `az1` needs the default backend `az0` (central) and `az1` (the local cluster), so the other backends can be removed

fultonj replied:
I agree

fultonj commented on examples/dt/dcn/README.md:
```diff
@@ -0,0 +1,42 @@
+    - Cinder Backup using RBD for backend
+    - Glance using Multi Store Support and RBD for backend
+    - Nova using RBD for ephemeral storage
+    - Manila using CephFS for backend
```
Let's remove manila for now

fultonj commented:
Please add the following to the commit message and the github PR description.

`Depends-On: openstack-k8s-operators/ci-framework#2458`

fultonj commented on examples/dt/dcn/dataplane-post-ceph.md:
````diff
@@ -0,0 +1,72 @@
+Add compute host to aggregate. To add a compute host to the aggregate, use the following command:
+```bash
+oc rsh openstackclient openstack aggregate add host az0 edpm-compute-0.ctlplane.example.com
+```
````
## Additional Availability Zones

The above will result in `az0` with 3 compute nodes and one ceph cluster being deployed. To deploy additional AZs, e.g. `az1` and `az2`, each with 3 more compute nodes and one more ceph cluster, create updated values files and re-apply the kustomizations. An Ansible role which does this in our CI system for testing is available to be reviewed for details.

https://github.com/openstack-k8s-operators/ci-framework/tree/main/roles/ci_dcn_site

sbekkerm replied:
done

fultonj commented:
> > @sbekkerm @fultonj One more question, I am not sure what the final decision was but do we want to include multicell (cell per site) into this DCN DT? It's just adding some extra configuration for the control plane which I can do.
> 
> @krcmarik do you want to try that in a stacked patch.
> 
> i.e. create a new PR which has been rebased on this one so that we can merge this one sooner.

We are using cells now. We can do follow ups for further changes.

fultonj commented:
> The README contains all the steps to reproduce the environment. Could you please clarify what specifically is unclear?
> 
> 1. These steps for deploying the control plane:
>    https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/control-plane.md
> 2. These steps for preparing nodes for ceph installation:
>    https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-pre-ceph.md
> 3. These steps for configuring nodes after ceph installation:
>    https://github.com/sbekkerm/architecture/blob/dcn/examples/va/dcn/dataplane-post-ceph.md
> 
> The Jinja templates are not involved in the deployment process, as they are only used by CI to set the "CHANGEME" parameters in values.yaml and service-valiues.yaml

Yes, all of these concerns have since been addressed.

fultonj commented on examples/dt/dcn/README.md:
```diff
@@ -0,0 +1,43 @@
+# Distributed Compute Node (DCN) OpenStack Architecture with HCI and Ceph
+
```
Please add the following here:
```
**Based on OpenStack K8S operators from the "main" branch of the [OpenStack Operator repo](https://github.com/openstack-k8s-operators/openstack-operator/commit/ec7210b825a3c355007d1f1fc11a2952ba4a9262) on Oct 22nd, 2024**
```

fultonj commented:
> @sbekkerm I see two changes needed up front.
> 
> 1. Written Instructions
> 2. VA vs DT

These changes have been addressesd.

fmount commented on examples/dt/dcn/service-values.yaml:
```diff
@@ -0,0 +1,237 @@
+        rbd_cluster_name = az2
+        backend_availability_zone = az2
+  glance:
+    customServiceConfig: |
```
This `customServiceConfig` at `glance` level seems redundant to me.
Given you're defining the backend for each API under `glanceAPIs` (including az0), you can basically remove from L69 to L79 [1].
You  inherit the top-level `customServiceConfig` only if it's not defined for that specific instance, and this does not seem the case.
@fultonj do you want to update the glance config? 

[1] https://github.com/openstack-k8s-operators/glance-operator/blob/main/controllers/glance_controller.go#L890

fmount replied:
Technically it doesn't block your deployment because as per the link I mentioned earlier the right config ends up being on the right instance, but I think this is simply redundant and should be patched.

fultonj replied:
I agree but per our discussion I will submit a follow up patch to address this.

fultonj reviewed (commented):
I'm +2 on merging this.

fmount reviewed (approved):
/lgtm

abays reviewed (approved):
/lgtm
/approve

fultonj reviewed (approved):
/approve
/lgtm

fultonj commented:
> > > @sbekkerm @fultonj One more question, I am not sure what the final decision was but do we want to include multicell (cell per site) into this DCN DT? It's just adding some extra configuration for the control plane which I can do.
> > 
> > 
> > @krcmarik do you want to try that in a stacked patch.
> > i.e. create a new PR which has been rebased on this one so that we can merge this one sooner.
> 
> We are using cells now. We can do follow ups for further changes.

To clarify, we have cells, but we're not using cells per site. I think we should address that in a follow up patch.
